### ISSUE TITLE
Add audio

### ISSUE THREAD

**Issue by picsi** [birtles/rikaichamp#42]
Somehow this Rikaisama port (https://forum.koohii.com/thread-15119.html) has audio included, played when pressing F. It would be amazing to have that on rikaichamp.

**Comment by Garethp** [comment]
If you do want to implement it easily, [here's](https://github.com/Garethp/RikaiRebuilt/blob/master/src/audioPlayer.js) where the code for the AudioPlayer is. May need a tad adjusting, but overall it's pretty solid. That class should work (mostly) standalone, and comes with detection for the clip they play when there's no audio

**Comment by birtles** [comment]
Thanks again Gareth!

That looks like an audio plugin I wrote for Anki once! (I think I discovered there are actually two hashes for not found audio). If you don't mind I'll gladly steal your code! (And convert it to TypeScript, and add tests I guess).

**Comment by AOEChamp** [comment]
Bump, this would be a really useful feature. I can submit a PR if that helps.

I'm thinking:
- Audio icon in the popup
- Setting to show icon and maybe another for autoplay
- Playback code itself is pretty straightforward

**Comment by birtles** [comment]
Hi! A PR would be great, thank you!

For what it's worth, I hadn't forgotten about this but I had in mind to address it after converting the popup to Preact since it will be easier to add interactivity then. I think there's no reason to wait for that, however.

One issue, however, is we need to make sure we are legally allowed to use the audio files. I think in the past everyone was using the Japanesepod101 files? Is that right? If so I want to be sure they are ok with us linking to their files.

**Comment by Garethp** [comment]
If you do what I did and try to add the audio playing to the background script you might run into some issue. Manifest V3 doesn't allow that since it runs as a service worker and the `Audio` component is DOM based though. That being said, if you're playing it in the pop-up through interaction you're good to go! I'm keeping mine in the background though so that I don't have to refactor it and instead I'm just opening an offscreen document with and sending it a message in Manifest V3

AudioPlayer.ts
```ts
  static async play(entry, config) {
    this.getAudioBase64(entry).then((content) => {
      const audioSource = "data:audio/wav;base64," + content;
      const audioVolume = config.audioVolume / 100 || 1;

      if (isManifestV3) return this.playOffscreen(audioSource, audioVolume);

      const audio = new Audio(audioSource);
      audio.volume = audioVolume;
      audio.play();
    });
  }

  static async playOffscreen(source: string, volume: number) {
    await this.createOffscreen();
    await browser.runtime.sendMessage({ playAudio: { source, volume } });
  }

  static async createOffscreen() {
    // @ts-ignore
    if (await chrome.offscreen.hasDocument()) return;
    // @ts-ignore
    await chrome.offscreen.createDocument({
      url: "../offscreen-audio/index.html",
      reasons: ["AUDIO_PLAYBACK"],
      justification: "User Requested Audio Playback",
    });
  }
```

offscreen-audio/index.html
```html
<script src="../build/offscreen-audio.js"></script>
```

OffscreenAudio.ts
```ts
// @ts-ignore
const browser = require("webextension-polyfill");

browser.runtime.onMessage.addListener((msg) => {
  if ("playAudio" in msg) playAudio(msg.playAudio);
});

// Play sound with access to DOM APIs
function playAudio({ source, volume }) {
  const audio = new Audio(source);
  audio.volume = volume;
  audio.play();
}
```

**Comment by birtles** [comment]
Yes, good idea about using the background page. It's certainly simpler if we can play audio from the content script. We'll need to make sure host pages with restrictive CSP policies don't prevent that, however.

For example, if the page the user is looking up text on is served with `Content-Security-Policy: media-src none;` or `Content-Security-Policy: media-src self;` it's possible that it might prevent an audio element we add to the DOM from fetching the audio file. If we create the audio element on-demand using the `Audio` constructor in the isolated world, however, we might be ok. We'd need to test, I think.

If CSP ends up being a problem we'll likely need to do what @Garethp suggested and play the audio using an offscreen document in Chromium browsers and in the background page in other browsers.

**Comment by birtles** [comment]
But like I previously mentioned, I think the first step is to check if languagepod101 is OK with using their content.

**Comment by AOEChamp** [comment]
> But like I previously mentioned, I think the first step is to check if languagepod101 is OK with using their content.

Are you going to contact them?

Tbh, to me it seems like one of those better to ask for forgiveness scenarios. Other extensions use their audio too, and I'd imagine they would have some sort of authentication if they really cared: https://github.com/FooSoft/yomichan/issues/706

**Comment by birtles** [comment]
> Are you going to contact them?

Yes, I've sent them a mail now. I'll update this issue when I hear back from them.

**Comment by xrishox** [comment]
I think that a better long term sustainable approach would be to add support for local audio server.

https://github.com/themoeway/local-audio-yomichan

it makes it portable so that it can be hosted anywhere and if one plan goes down someone else can install it. if forvo/jpod/etc go away it still works. it's also fully self hostable and works entirely offline. there is also files already made for all of forvo, all of jpod, all of nhk, and all of shinmeikai. yomitan has already built all of the infrastructure for all of this to work, plus it will allow you to plug in to the online formats they already have.

for people who don't want to host it themself 

https://github.com/jamesnicolas/yomichan-forvo-server

this is a simple version that doesn't require you to self host it, which pulls directly from forvo. because in both instances the audio fetching and searching is done by the anki plugin it would drastically reduce the amount of work it would take to implement it.

**Comment by birtles** [comment]
Thanks for those links! I didn't know about those projects.

The main issue is really just getting permission to use those files. I connected with one of the founders of Japanesepod101 on LinkedIn but I haven't been able to get a response from him.

I'll make a note to have another try tomorrow.

I'm trying to make 10ten "just work", particularly for non-technical users, so I'd like to avoid requiring users to download extra files, run servers, install Anki etc. as far as possible. If I can get permission to use the files, I might be able to host them too to provide even better performance.

**Comment by Anthonyy232** [comment]
Hi, not to be an annoying prick too much, but any updates on this? I think TTS would be very helpful as a sanity check. Thank you for your work on this project.

**Comment by Anthonyy232** [comment]
Did a quick and dirty approach [here](https://github.com/Anthonyy232/10ten-tts-reader) if anyone wants to have something to hold them over until official implementation. Not going to do a PR because it's very likely not the best way to implement it, but it *does* work if it will help.

**Comment by birtles** [comment]
Thanks, that's awesome. Sorry for the late reply.

@enellis and I started sketching out a solution for this a while back. I think we want to do things a little differently—tie the playback to the reading more closely, pre-generate audio samples, and we can't use japanesepod etc. But what you have is definitely a great reference point.

@enellis any chance you might have cycles to work on this in the near future?

**Comment by enellis** [comment]
> [@enellis](https://github.com/enellis) any chance you might have cycles to work on this in the near future?

Yes, I'll definitely look more into this in the coming weeks.
